### PR TITLE
feat(crates-io): publish gtest on crates.io

### LINKED ISSUE #3501: Figure out a solution for publishing `gtest`
### Problem to Solve

`gtest` contains `core-processor` as dependency which is hard to publish to crates-io

### Possible Solution

1. fork the unpublished dependency ( `sp-allocator`, ... )
2. replace the unpublished dependency on publishing

### Notes

_No response_

### PR DESCRIPTION
**Pull request by clearloop**
> [!TIP]
[gtest-v1.0.4](https://crates.io/crates/gtest) is now available on crates.io based on c815950

--- 

Resolves #3501 

- [x] publish `gear-sandbox-*`
  - [x] publish `-env`
  - [x] publish `-host`
  - [x] publish `sandbox`
- [x] publish `gear-lazy-pages`
- [x] publish `gear-core-backend`
- [x] publish `core-processor`
- [x] publish `gtest`

Significant changes have been split into 

- #3623 
- #3626
- #3629

Merging this PR could take some time, bcz we are about to upgrade to `substrate-v1.0.0` in `v1.0.4` (#3608) which will break part of the logic in the current implementation immediately, see [the patch handler of substrate](https://github.com/gear-tech/gear/blob/cl/issue-3501/utils/crates-io/src/handler.rs#L47) for the details

continue this PR after `v1.0.4`

---

see the tip on the top of this comment

@gear-tech/dev

### CODE REVIEW

**Comment by clearloop** [comment]
now have bugs publishing `sandbox-*`, because the versioning of sp-wasm-interface is messed up since we have modified a log in our fork,  https://github.com/gear-tech/gear/pull/3603 cleans it in sandbox

**Comment by breathx** [comment]
v1.0.4 is in master

File: utils/crates-io/src/version.rs [hunk 1]
```diff
@@ -33,17 +33,26 @@ struct Version {
 }
 
 /// Verify if the package has already been published.
-pub fn verify(name: &str, version: &str) -> Result<bool> {
+pub fn verify(mut name: &str, version: &str) -> Result<bool> {
+    println!("Verifying {}@{} ...", &name, &version);
+
+    if name == "gear-core-processor" {
```
Comment (breathx) [inline utils/crates-io/src/version.rs:39]: maybe go through all aliases finding match? here and in other places

File: utils/crates-io/src/handler.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    for (package, alias) in crate::PACKAGE_ALIAS {
+        if index.contains(&package) {
+            index.push(alias);
+        }
+    }
+}
+
+/// Patch the workspace manifest.
+pub fn patch_workspace(name: &str, table: &mut toml_edit::InlineTable) {
+    match name {
```
Comment (breathx) [inline utils/crates-io/src/handler.rs:55]: runtime interface?

File: utils/crates-io/src/handler.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+/// gear-core-processor handler.
+mod core_processor {
+    use toml_edit::{Document, InlineTable};
+
+    /// Rename core processor related package in the
+    /// manifest of workspace since `gear-core-processor`
+    /// has been published by others.
+    pub fn patch_workspace(name: &str, table: &mut InlineTable) {
+        match name {
```
Comment (breathx) [inline utils/crates-io/src/handler.rs:85]: why do we remove if exist and add if not?

**Review by breathx — APPROVED** [review]

File: utils/crates-io/src/version.rs (hunk 1 above)
Comment (clearloop) [reply utils/crates-io/src/version.rs:39]: We have all of our packages names on crates-io except `gear-core-processor`, see https://crates.io/crates/gear-core-processor, it is taken by others, I'll gather this func in `handler.rs` and gather the docs of `core-processor`

File: utils/crates-io/src/handler.rs (hunk 2 above)
Comment (clearloop) [reply utils/crates-io/src/handler.rs:55]: `sp-runtime-interface` is included in the substrate dependencies, for the modification of `gear-runtime-interface`, it is included in `handler::patch`

File: utils/crates-io/src/handler.rs (hunk 3 above)
Comment (clearloop) [reply utils/crates-io/src/handler.rs:96]: it's for pointing the both alias to https://crates.io/crates/core-processor, see a58644f

**Review by ukint-vs — APPROVED** [review]

**Comment by breathx** [comment]
kindly asking @shamilsan for review
and if it's done from your side @clearloop feel free to merge

File: gtest/Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,7 @@
 [package]
 name = "gtest"
-version = "0.1.0"
+description = "Testing utils for developing gear programs."
```
Comment (shamilsan) [inline gtest/Cargo.toml:3]:
```suggestion
description = "Testing utils for developing Gear programs."
```

**Review by shamilsan — APPROVED** [review]
Looks good to me

**Pull request merged**